metrics_sdk/lib/opentelemetry/sdk/metrics/meter_provider.rb

```diff
@@ -0,0 +1,121 @@
+        #
+        # Calls MetricReader#shutdown for all registered MetricReaders.
+        # Calls MetricExporter#shutdown for all registered MetricExporters.
+        #
+        # After this is called all the newly created {Meter}s will be no-op.
+        #
+        # @param [optional Numeric] timeout An optional timeout in seconds.
+        # @return [Integer] Export::SUCCESS if no error occurred, Export::FAILURE if
+        #   a non-specific failure occurred, Export::TIMEOUT if a timeout occurred.
+        def shutdown(timeout: nil)
+          @mutex.synchronize do
+            if @stopped
+              OpenTelemetry.logger.warn('calling MetricProvider#shutdown multiple times.')
+              Export::FAILURE
+            else
+              start_time = OpenTelemetry::Common::Utilities.timeout_timestamp
+
+              results = @metric_readers.map do |metric_reader|
+                remaining_timeout = OpenTelemetry::Common::Utilities.maybe_timeout(timeout, start_time)
+                break [Export::TIMEOUT] if remaining_timeout&.zero?
+                metric_reader.shutdown(timeout: remaining_timeout)
+              end
+
+              @stopped = true
+              results.max || Export::SUCCESS
+            end
+          end
+        end
+
+        # This method provides a way for provider to notify the registered
+        # {MetricReader} instances, so they can do as much as they could to consume
+        # or send the metrics. Note: unlike Push Metric Exporter which can send data on
+        # its own schedule, Pull Metric Exporter can only send the data when it is
+        # being asked by the scraper, so ForceFlush would not make much sense.
+        #
+        # @param [optional Numeric] timeout An optional timeout in seconds.
+        # @return [Integer] Export::SUCCESS if no error occurred, Export::FAILURE if
+        #   a non-specific failure occurred, Export::TIMEOUT if a timeout occurred.
+        def force_flush(timeout: nil)
+          @mutex.synchronize do
+            return Export::SUCCESS if @stopped
+
+            start_time = OpenTelemetry::Common::Utilities.timeout_timestamp
+            results = @metric_readers.map do |metric_reader|
+              remaining_timeout = OpenTelemetry::Common::Utilities.maybe_timeout(timeout, start_time)
+              return Export::TIMEOUT if remaining_timeout&.zero?
+
+              metric_reader.force_flush(timeout: remaining_timeout)
+            end
+            results.max || Export::SUCCESS
+          end
+        end
+
+        # Adds a new MetricReader to this {MeterProvider}.
+        #
+        # @param metric_reader the new MetricReader to be added.
+        def add_metric_reader(metric_reader)
+          @mutex.synchronize do
+            if @stopped
+              OpenTelemetry.logger.warn('calling MetricProvider#add_metric_reader after shutdown.')
+              return
+            end
+
+            @metric_readers = @metric_readers.push(metric_reader)
```

Comment:
Don't really need the assignment here, since `push` mutates the receiver.
```suggestion
            @metric_readers.push(metric_reader)
```
Dumb question: this will return the `@metric_readers`, right? Is that a good idea or should we explicitly return `nil`? I know we do the same in `TracerProvider` - I think that might be a mistake as well.